### PR TITLE
Documentation facelift

### LINKED ISSUE #226: Rewrite documentation site
[Astro Starlight](https://starlight.astro.build/) looks really great for this. I've just rewritten my personal website with Astro and am very impressed.

However, Starlight doesn't have an automatic docstring converter so an extra tool like `mkdocstrings-python` is needed to generate `.md` files.

`mkdocs` can also be used and is a staple in the Python community. However, it's old and less hyped. As someone who sees this as an opportunity to learn new stuff, things like `mkdocs`, `sphinx` are not very appealing.

### PR DESCRIPTION
**Pull request by wiseodd**
Closes #226.

I decided to use `mkdocs-material`, which is the standard in Python (see Astral's docs etc.).

@aleximmer, I added a deploy workflow. Could you disable the current one for `pdoc`? (Seems like it's hidden in the repo settings page.) 

To review:

```bash
uv sync --all-extras
uv run mkdocs serve
```

then open `localhost:8000`

### CODE REVIEW

File: docs/devs_guide.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+**With `uv`**
+
+```bash
+# assuming the repository was cloned
+uv sync --all-extras
+# create docs and write to html
+uv run bash update_docs.sh
+# .. or serve the docs directly
+uv run pdoc --http 0.0.0.0:8080 laplace --template-dir template
```
Comment (aleximmer) [inline docs/devs_guide.md:63]: Should this be updated?

File: docs/devs_guide.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+**With `pip`**
+
+```bash
+# assuming the repository was cloned
+pip install -e ".[dev]"
+# create docs and write to html
+bash update_docs.sh
+# .. or serve the docs directly
+pdoc --http 0.0.0.0:8080 laplace --template-dir template
```
Comment (aleximmer) [inline docs/devs_guide.md:74]: Same here potentially

**Review by aleximmer — APPROVED** [review]
Thanks a lot for updating the docs. They look super good now 💯 

I got some warnings when running `uv run mkdocs serve`, not sure if these are serious:
```
SyntaxWarning: invalid escape sequence '\i'
             File "Laplace/.venv/lib/python3.12/site-packages/_griffe/agents/visitor.py", line 203, in get_module
               top_node = compile(self.code, mode="exec", filename=str(self.filepath), flags=ast.PyCF_ONLY_AST, optimize=1)
             File "Laplace/laplace/lllaplace.py", line 510, in
               """Here not much changes in terms of GP inference compared to FunctionalLaplace class.
```

It also shows us some docstring parameter descriptions that we should fix:
```
WARNING -  griffe: laplace/baselaplace.py:1957: Parameter 'num_data' does not appear in the function signature
WARNING -  griffe: laplace/baselaplace.py:1957: Parameter 'diagonal_kernel' does not appear in the function signature
WARNING -  griffe: laplace/baselaplace.py:1957: No types or annotations for parameters ['See']
WARNING -  griffe: laplace/baselaplace.py:1957: Parameter 'See' does not appear in the function signature
WARNING -  griffe: laplace/utils/matrix.py:296: Parameter 'dampen' does not appear in the function signature
WARNING -  griffe: laplace/utils/subnetmask.py:361: No types or annotations for parameters ['parameter_names']
WARNING -  griffe: laplace/utils/subnetmask.py:361: Parameter 'parameter_names' does not appear in the function signature
```
Should we open an issue to fix it after the PR?

For deploying the docs, I will disable the old docs once you merge.

**Review by runame — APPROVED** [review]
Haven't looked at this in detail yet, but looks good! Feel free to just merge.

**Comment by wiseodd** [comment]
@aleximmer fixed the issues. For the improper docstrings, I agree, let's open an issue for this. I also want to refactor the codebase to make it more extendible.

Feel free to merge and test whether the new docs site is properly deployed.

**Pull request merged**